presto-spi/src/main/java/io/prestosql/spi/block/LazyBlock.java

```diff
@@ -159,131 +156,270 @@ public int compareTo(int leftPosition, int leftOffset, int leftLength, Block rig
     }
 
     @Override
     public Block copyRegion(int position, int length)
     {
-        assureLoaded();
-        return block.copyRegion(position, length);
+        return getBlock().copyRegion(position, length);
     }
 
     @Override
     public boolean isNull(int position)
     {
-        assureLoaded();
-        return block.isNull(position);
+        return getBlock().isNull(position);
+    }
+
+    @Override
+    public final List<Block> getChildren()
+    {
+        return singletonList(getBlock());
     }
 
     public Block getBlock()
     {
-        assureLoaded();
-        return block;
+        return lazyData.getBlock();
     }
 
     @Override
     public boolean isLoaded()
     {
-        return block != null && block.isLoaded();
+        return lazyData.isLoaded();
     }
 
     @Override
     public Block getLoadedBlock()
     {
-        assureLoaded();
-        block = block.getLoadedBlock();
-        return block;
+        return lazyData.getFullyLoadedBlock();
+    }
+
+    public static void listenForLoads(Block block, Consumer<Block> listener)
+    {
+        requireNonNull(block, "block is null");
+        requireNonNull(listener, "listener is null");
+
+        LazyData.addListenersRecursive(block, singletonList(listener));
+    }
+
+    private static class RegionLazyBlockLoader
+            implements LazyBlockLoader
+    {
+        private final LazyData delegate;
+        private final int positionOffset;
+        private final int length;
+
+        public RegionLazyBlockLoader(LazyData delegate, int positionOffset, int length)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+            this.positionOffset = positionOffset;
+            this.length = length;
+        }
+
+        @Override
+        public Block load()
+        {
+            return delegate.getBlock().getRegion(positionOffset, length);
+        }
     }
 
-    private void assureLoaded()
+    private static class PositionLazyBlockLoader
+            implements LazyBlockLoader
     {
-        if (block != null) {
-            return;
+        private final LazyData delegate;
+        private final int[] positions;
+        private final int offset;
+        private final int length;
+
+        public PositionLazyBlockLoader(LazyData delegate, int[] positions, int offset, int length)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+            this.positions = requireNonNull(positions, "positions is null");
+            this.offset = offset;
+            this.length = length;
         }
-        block = requireNonNull(loader.load(), "loader returned null");
 
-        while (block instanceof LazyBlock) {
-            block = ((LazyBlock) block).getBlock();
+        @Override
+        public Block load()
+        {
+            return delegate.getBlock().getPositions(positions, offset, length);
+        }
+    }
+
+    private static class LazyData
```

Comment:
the `Block` contract is a bit unclear if blocks are immutable. For example `io.prestosql.spi.block.DictionaryBlock` has `volatile` fields, which would suggest that blocks can be accessed from multiple threads. I think for `LazyBlock` if would be very problematic.